lightning/src/util/chacha20poly1305rfc.rs

```diff
@@ -97,6 +120,59 @@ mod real_chachapoly {
 #[cfg(not(fuzzing))]
 pub use self::real_chachapoly::ChaCha20Poly1305RFC;
 
+/// Enables simultaneously writing and encrypting a byte stream into a Writer.
+struct ChaChaPolyWriter<'a, W: Writer> {
+	pub chacha: &'a mut ChaCha20Poly1305RFC,
+	pub write: &'a mut W,
+}
+
+impl<'a, W: Writer> Writer for ChaChaPolyWriter<'a, W> {
+	// Encrypt then write bytes from `src` into Self::write.
+	// `ChaCha20Poly1305RFC::finish_and_get_tag` can be called to retrieve the tag after all writes
+	// complete.
+	fn write_all(&mut self, src: &[u8]) -> Result<(), io::Error> {
+		let mut src_idx = 0;
+		let mut write_buffer = [0; 8192];
+		let mut bytes_written = (&mut write_buffer[..]).write(&src[src_idx..]).expect("In-memory writes can't fail");
+		while bytes_written != 0 {
```

Comment:
nit: Could make this a `loop` with a `break` and move the statements inside to avoid some duplication.

Reply:
Let me know what you think of the latest version. I implemented it per Matt's previous comment: https://github.com/lightningdevkit/rust-lightning/pull/1518#discussion_r900284301